src/common.ts

```diff
@@ -54,4 +55,57 @@ export class Common {
     logger.debug(`loginUrl: ${loginUrl}`);
     return loginUrl;
   }
+
+  public static async identifyPossibleScratchOrgs(orgAuthInfo: AuthInfo): Promise<void> {
+    const fields = orgAuthInfo.getFields();
+    const logger = await Logger.child('Common', { tag: 'identifyPossibleScratchOrgs' });
+
+    // return if we already know the hub or we know it is a devhub or prod-like
+    if (fields.isDevHub || fields.devHubUsername) return;
+    // there are no hubs to ask, so quit early
+    if (!(await AuthInfo.hasAuthentications())) return;
+    logger.debug('getting devHubs from authfiles');
+
+    // TODO: return if url is not sandbox-like to avoid constantly asking about production orgs
+    // TODO: someday we make this easier by asking the org if it is a scratch org
+
+    const hubAuthInfos = await this.getDevHubAuthInfos();
+    logger.debug(`found ${hubAuthInfos.length} DevHubs`);
+    if (hubAuthInfos.length === 0) return;
+
+    // ask all those orgs if they know this orgId
+    await Promise.all(
+      hubAuthInfos.map(async (hubAuthInfo) => {
+        try {
+          const devHubOrg = await Org.create({ aliasOrUsername: hubAuthInfo.getUsername() });
+          const conn = devHubOrg.getConnection();
+          const data = await conn.query<QueryResult<{ Id: string }>>(
+            `select Id from ScratchOrgInfo where ScratchOrg = '${sfdc.trimTo15(orgAuthInfo.getFields().orgId)}'`
+          );
+          if (data.totalSize > 0) {
+            // if any return a result
+            logger.debug(`found orgId ${fields.orgId} in devhub ${hubAuthInfo.getUsername()}`);
+            try {
+              await orgAuthInfo.save({ ...orgAuthInfo.getFields(), devHubUsername: hubAuthInfo.getUsername() });
+              logger.debug(`set ${hubAuthInfo.getUsername()} as devhub for scratch org ${orgAuthInfo.getUsername()}`);
+            } catch (error) {
+              logger.debug(`error updating auth file for ${orgAuthInfo.getUsername()}`, error);
+            }
+          }
+        } catch (error) {
+          logger.error(`Error connecting to devhub ${hubAuthInfo.getUsername()}`, error);
+        }
+      })
+    );
+  }
+
+  public static async getDevHubAuthInfos(): Promise<AuthInfo[]> {
+    return (
+      await Promise.all(
+        (await AuthInfo.listAllAuthFiles())
+          .map((fileName) => basename(fileName, '.json'))
+          .map((username) => AuthInfo.create({ username }))
```

Comment:
@mshanemc I have noticed that when creating authinfos that core makes a call to native keychain impl to get secrets. Since dev hub attribute is not encrypted, how about fetching w/o requesting fields be decrypted?